### PR TITLE
Update support for nginx on local docker

### PR DESCRIPTION
**Pull request by jlewitt1**

### CODE REVIEW

File: runhouse/docker/slim/public-key-auth/Dockerfile [hunk 1]
```diff
@@ -41,7 +41,6 @@ RUN --mount=type=secret,id=ssh_key,dst=/tmp/id_rsa.pub \
 
 # Update SSHD Configuration to allow public key authentication and disable password and root login
 RUN echo "PasswordAuthentication no" >> /etc/ssh/sshd_config && \
-    echo "PermitRootLogin no" >> /etc/ssh/sshd_config && \
```
Comment (DenisYay) [inline runhouse/docker/slim/public-key-auth/Dockerfile:44]: Should we set "PermitRootLogin yes" to be explicit?

File: runhouse/docker/slim/public-key-auth/Dockerfile [hunk 2]
```diff
@@ -51,7 +50,10 @@ RUN echo "[supervisord]" > /etc/supervisor/conf.d/supervisord.conf && \
     echo "[program:sshd]" >> /etc/supervisor/conf.d/supervisord.conf && \
     echo "command=/usr/sbin/sshd -D" >> /etc/supervisor/conf.d/supervisord.conf && \
     echo "stdout_logfile=/var/log/sshd.log" >> /etc/supervisor/conf.d/supervisord.conf && \
-    echo "stderr_logfile=/var/log/sshd.err" >> /etc/supervisor/conf.d/supervisord.conf
+    echo "stderr_logfile=/var/log/sshd.err" >> /etc/supervisor/conf.d/supervisord.conf  && \
+    echo "[program:nginx]" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "command=nginx -g 'daemon off;'" >> /etc/supervisor/conf.d/supervisord.conf
```
Comment (DenisYay) [inline runhouse/docker/slim/public-key-auth/Dockerfile:55]: Do we need to install nginx?

File: runhouse/docker/slim/public-key-auth/Dockerfile [hunk 3]
```diff
@@ -51,7 +50,10 @@ RUN echo "[supervisord]" > /etc/supervisor/conf.d/supervisord.conf && \
     echo "[program:sshd]" >> /etc/supervisor/conf.d/supervisord.conf && \
     echo "command=/usr/sbin/sshd -D" >> /etc/supervisor/conf.d/supervisord.conf && \
     echo "stdout_logfile=/var/log/sshd.log" >> /etc/supervisor/conf.d/supervisord.conf && \
-    echo "stderr_logfile=/var/log/sshd.err" >> /etc/supervisor/conf.d/supervisord.conf
+    echo "stderr_logfile=/var/log/sshd.err" >> /etc/supervisor/conf.d/supervisord.conf  && \
+    echo "[program:nginx]" >> /etc/supervisor/conf.d/supervisord.conf && \
```
Comment (DenisYay) [inline runhouse/docker/slim/public-key-auth/Dockerfile:54]: Should starting nginx be guarded by an build-time ARG, similar to RUNHOUSE_VERSION?

File: runhouse/servers/http/http_server.py [hunk 4]
```diff
@@ -135,7 +135,7 @@ def __init__(
             @app.get("/spans")
             @validate_cluster_access
             def get_spans(request: Request):
-                print("Calling get_spans")
+                logger.info("Calling get_spans")
```
Comment (DenisYay) [inline runhouse/servers/http/http_server.py:138]: We can probably remove this line

File: runhouse/servers/nginx/config.py [hunk 5]
```diff
@@ -56,8 +56,14 @@ def configure(self):
             raise RuntimeError("Failed to configure Nginx")
 
     def reload(self):
+        # If running locally (e.g. in a docker container) systemctl may not be available
+        reload_cmd = (
+            "sudo service nginx start && sudo nginx -s reload"
+            if self.address == "localhost"
```
Comment (DenisYay) [inline runhouse/servers/nginx/config.py:62]: I wonder if we need a more robust check here.. do we need to check for 127.0.0.1 or 0.0.0.0 explicitly?

**Review by DenisYay — APPROVED** [review]

File: runhouse/docker/slim/public-key-auth/Dockerfile (hunk 2 above)
Comment (jlewitt1) [reply runhouse/docker/slim/public-key-auth/Dockerfile]: good point, I don't think we need to install here, should suffice to run subprocess commands on the server where needed

File: runhouse/servers/http/http_server.py (hunk 4 above)
Comment (DenisYay) [reply runhouse/servers/http/http_server.py]: I removed it in https://github.com/run-house/runhouse/pull/142

File: tests/test_http_server.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+@pytest.mark.dockertest
+@pytest.mark.httpservertest
+def test_public_key_cluster_is_up(local_docker_cluster_public_key):
+    cluster = local_docker_cluster_public_key
+    cluster.check_server()
+    assert cluster.is_up()  # Should be true for a Cluster object
+
+
+@pytest.mark.dockertest
```
Comment (DenisYay) [inline tests/test_http_server.py:23]: I've created a new file called [`test_telemetry.py`](https://github.com/run-house/runhouse/blob/d1993e3e8fce782fae5f1001bb7b5b5f32c07870/tests/test_telemetry.py#L9). This PR is already merged into main.
In the PR, I've moved the following method into `test_telemetry.py`:
`def test_public_key_cluster_has_telemetry(local_docker_cluster_telemetry_public_key):`

Should we move `def test_public_key_cluster_has_telemetry_with_tls(
    local_docker_cluster_telemetry_public_key_with_tls,` there too?

Also, let's annotate both tests with
`
@pytest.mark.dockertest
@pytest.mark.telemetrytest
`
Thanks!

File: tests/test_resources/test_clusters/conftest.py [hunk 7]
```diff
@@ -391,6 +391,119 @@ def local_docker_cluster_public_key(request):
...
+        host="localhost",
+        server_host="0.0.0.0",
+        server_connection_type=server_connection_type,
+        den_auth=den_auth,
+        ssh_port=local_ssh_port,
+        ssh_creds={
+            "ssh_user": SSH_USER,
+            "ssh_private_key": keypath,
+        },
+        enable_local_span_collection=True,
```
Comment (DenisYay) [inline tests/test_resources/test_clusters/conftest.py:437]: Please rename to `use_local_telemetry = True`

File: tests/test_resources/test_clusters/conftest.py [hunk 8]
```diff
@@ -391,6 +391,119 @@ def local_docker_cluster_public_key(request):
         client.images.prune()
 
 
+def build_local_docker_cluster_with_telemetry(
```
Comment (DenisYay) [inline tests/test_resources/test_clusters/conftest.py:394]: We probably need another round of merging the changes in this file with the changes from main:

https://github.com/run-house/runhouse/blob/d1993e3e8fce782fae5f1001bb7b5b5f32c07870/tests/test_resources/test_clusters/conftest.py#L335

File: tests/test_resources/test_clusters/conftest.py [hunk 9]
```diff
@@ -391,6 +391,119 @@ def local_docker_cluster_public_key(request):
...
+    image_name = "keypair-telemetry"
+    dir_name = "public-key-auth"
+    keypath = str(
+        Path(
+            rh.configs.get("default_keypair", "~/.ssh/runhouse/docker/id_rsa")
+        ).expanduser()
+    )
+    local_ssh_port = BASE_LOCAL_SSH_PORT + 6
+
+    port_fwds = [f"{local_ssh_port}:22"]
```
Comment (DenisYay) [inline tests/test_resources/test_clusters/conftest.py:411]: Should be `port_fwds=[f"{local_ssh_port}:22", "32300:32300"],`

**Review by DenisYay — APPROVED** [review]

File: runhouse/resources/hardware/cluster.py [hunk 10]
```diff
@@ -579,10 +579,17 @@ def _use_https(self) -> bool:
...
     @property
     def _use_nginx(self) -> bool:
-        """Use Nginx if the server port is set to the default HTTP (80) or HTTPS (443) port.
+        """Use Nginx if the server port is set to the default HTTP (80/ 8080) or HTTPS (443 / 8443) port.
         Note: Nginx will serve as a reverse proxy, forwarding traffic from the server port to the Runhouse API
         server running on port 32300."""
-        return self.server_port in [self.DEFAULT_HTTP_PORT, self.DEFAULT_HTTPS_PORT]
+        # Support 8080 and 8443 as defaults for running locally (e.g. in a docker container, to avoid port conflicts
+        # on the host)
+        return self.server_port in [
```
Comment (jlewitt1) [inline runhouse/resources/hardware/cluster.py:587]: @dongreenberg wdyt about adding 8080 http and 8443 https as "default" ports for nginx? useful when running locally in docker

**Pull request closed without merging**